File: java/leetcode/LongestUncommon.java
//Leetcode problem 521 Longest Uncommon Subsequence I
//Leetcode problem 522 Longest Uncommon Subsequence II
//Solution written by Xuqiang Fang on 28 May, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

class Solution{
    public int findLUSlength(String a, String b){
        if(a.length() != b.length())
            return Math.max(a.length(), b.length());
        else if(a.equals(b))
            return -1;
        return a.length();
    }	

    //well, sometimes the stupidest idea works
    public int findLUSlength(String[] strs) {
        Map<String, Integer>[] s = new HashMap[11];    
        for(String t : strs){
            if(s[t.length()] == null){
                s[t.length()] = new HashMap<>();
            }
            s[t.length()].put(t, s[t.length()].getOrDefault(t,0)+1);
        }
        for(int i=10; i>=0; --i){
            if(s[i] != null){
                for(String k : s[i].keySet()){
                    if(s[i].get(k) == 1){// it is unique among all strings with the same length
                        if(i == 10){
                            return i;
                        }
                        boolean isSub = false;
                        for(int j=i+1; j<=10; ++j){
                            if(s[j] != null){
                                for(String l : s[j].keySet()){
                                    if(subsequence(k, l)){
                                        isSub = true;
                                        break;
                                    }
                                }
                                if(isSub){
                                    break;
                                }
                            }
                        }
                        if(!isSub){
                            return i;
                        }
                    }
                }
            }
        }
        return -1;
    }

    private boolean subsequence(String K, String L){
        if(K.length() == 0){
            return true;
        }
        int k = 0, l = 0;
        for(int i=0; i<Math.max(K.length(), L.length()); ++i){
            if(K.charAt(k) == L.charAt(l)){
                k++;
                if(k == K.length()){
                    return true;
                }
            }
            l++;
        }
        return false;
    }
}

public class LongestUncommon{
	public static void main(String[] args){
		Solution s = new Solution();
        //System.out.println(s.subsequence("", "abcddefgs"));
        //System.out.println(s.subsequence("abcd", "abcddefgs"));
        //System.out.println(s.subsequence("ddefgss", "abcddefgs"));
        String[] strs = {"aba", "cdc", "eae"};
        String[] str = {"aaa","aaa","aa"};
        System.out.println(s.findLUSlength(strs));
        System.out.println(s.findLUSlength(str));
	}
}

File: cpp/leetcode/LongestUncommon.cpp
//Leetcode Problem No  522 Longest Uncommon Subsequence II
//Solution written by Xuqiang Fang on 28 May, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    int findLUSlength(vector<string>& strs) {
        vector<unordered_map<string, int>> m(11, unordered_map<string, int>());    
        for(auto s : strs){
            if(m[s.length()].count(s) == 0){
                m[s.length()].insert({s, 1});
            }
            else{
                m[s.length()][s] += 1;
            }
        }
        for(int i=10; i>=0; --i){
            for(auto iter = m[i].begin(); iter != m[i].end(); ++iter){
                if(m[i][iter->first] == 1){
                    if(i == 10){
                        return i;
                    }
                    else{
                        bool isSub = false;
                        for(int j=i+1; j<=10; ++j){
                            for(auto it = m[j].begin(); it != m[j].end(); ++it){
                                if(sub(iter->first, it->first)){
                                    isSub = true;
                                    break;
                                }
                            }
                            if(isSub){
                                break;
                            }
                        }
                        if(!isSub){
                            return i;
                        }
                    }
                } 
            }
        }
        return -1;
    }
private:
    int sub(string K, string L){
        if(K.length() == 0){
            return 1;
        }
        int k = 0, l = 0;
        for(int i=0; i<max(K.length(), L.length()); ++i){
            if(K[k] == L[l]){
                k++;
                if(k == K.length()){
                    return 1;
                }
            }
            l++;
        }
        return 0;
    }
};

int main(){
    Solution s;
    vector<string> strs{"aba", "cdc", "eae"};
    cout << s.findLUSlength(strs) << endl;
    return 0;
}
